timely_beliefs/beliefs/classes.py

```diff
@@ -1467,6 +1492,124 @@ def resample_events(
 
         return df
 
+    @hybrid_method
+    def form_beliefs(
+        self,
+        belief_time: datetime,
+        source: BeliefSource,
+        event_start: datetime = None,
+        event_time_window: Optional[Tuple[datetime, datetime]] = (
+            None,
+            None,
+        ),
+        forecaster: Optional[BaseForecaster] = None,
+        inplace: bool = False,
+    ):
+        """Form new beliefs by applying a given forecaster.
+
+        :param belief_time:       Time at which the forecasts where made
+                                  (any belief after this time will be inaccessible to the forecaster).
+        :param source:            Source to assign the newly formed beliefs to.
+        :param event_start:       Set this to forecast a single event with the given start time.
+        :param event_time_window: Set this to forecast all events within the given time window.
+        :param forecaster:        Forecasting model. Currently, only sktime models are supported.
+                                  The default forecaster simply repeats the last known value.
+                                  Hint: set a seasonal periodicity to obtain a more reasonable (baseline) forecast.
+                                  For example:
+                                      >>> periodicity = pd.Timedelta("PT1W")
+                                      >>> forecaster = NaiveForecaster(sp=periodicity // df.event_resolution)
+        :param inplace:           If True, the new beliefs are added to the BeliefsDataFrame in place.
+                                  If False, the new beliefs are returned (use pd.concat to add them yourself).
+        """
+        if forecaster is None:
+            forecaster = NaiveForecaster(strategy="last")
+        if event_start is not None:
+            if event_time_window != (None, None):
+                raise ValueError(
+                    "Cannot pass both an event start and event start window."
+                )
+            event_time_window = (event_start, event_start + self.event_resolution)
+
+        # Check assumptions
+        if self.sensor.knowledge_time(event_time_window[0]) < belief_time:
+            # No backcasting
+            raise NotImplementedError("Backcasting is not implemented.")
+        if self.empty:
+            # At least 1 belief
+            raise NotImplementedError(
+                "Cannot form new beliefs given an empty data frame."
+            )
+        if self.lineage.number_of_sources != 1:
+            # Single source only
+            raise NotImplementedError(
+                "Cannot form new beliefs given multi-sourced data. Consider picking 1 source, e.g. using: df = df[df.index.get_level_values('source') == df.lineage.sources[0]]"
```

Comment:
Why is this not supported? Technical reasons?
Also, we could accept a source to narrow down what data to use first, right?

Reply:
> Why is this not supported? Technical reasons?

Missing implementation, for which I'm still unsure which direction to take it. If two sources have different expectations about the last value, what should the forecaster return? We don't yet have a feature to compute aggregate beliefs. For example, consider this input data:
```
                                                                                     event_value
event_start               belief_time               source   cumulative_probability             
2000-01-03 09:00:00+00:00 2000-01-01 01:00:00+00:00 Source A 0.5                           100.0
                                                    Source B 0.5                            25.0
```
Given equal weights, the input data could become:
```
                                                                                       event_value
event_start               belief_time               source     cumulative_probability             
2000-01-03 09:00:00+00:00 2000-01-01 01:00:00+00:00 Source A+B 0.5                            62.5
```
and the forecast could be `62.5` with `Source f(A+B)`. Alternatively, the forecaster could return:
- `[25.0, 100.0]` with separate sources `[Source f(A), Source f(B)]`.
- `[25.0, 62.5, 100.0]` with `Source f(A+B, d_A, d_B)` and CP values depending on assumptions on the underlying distributions `d` (e.g. CP values of `[0.5, 0.5, 0.5]`, `[0.33, 0.5, 0.67]` or `[0.25, 0.5, 0.75]`).
- `[25.0, 80.0, 100.0]` with `Source f(A+B, w)`given some skewed weight distribution `w` over the sources that favours source A.

> Also, we could accept a source to narrow down what data to use first, right?

We could. But imo that should not be the task of this function.

Reply:
>  But imo that should not be the task of this function.

Agreed.